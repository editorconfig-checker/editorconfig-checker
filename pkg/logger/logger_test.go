package logger

import (
	"bytes"
	"os"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
)

// interface for the closures passed to snapHelper
type loggerTest func()

// a helper function, that runs a closure with a logger, but redirects the Logger to write into a buffer, and then does snapshot testing on the buffer content
func snapHelper(t *testing.T, logger *Logger, test loggerTest) {
	t.Helper()
	buffer := bytes.Buffer{}
	logger.SetWriter(&buffer)
	test()
	snaps.MatchSnapshot(t, buffer.String())
}

func TestLoggerDebug(t *testing.T) {
	logger := GetLogger()

	snapHelper(t, &logger, func() {
		logger.Debug("this text should not be printed by a logger with default config")
	})

	snapHelper(t, &logger, func() {
		logger.DebugEnabled = true
		logger.Debug("this text should be printed when debug was enabled")
	})
}

func TestLoggerVerbose(t *testing.T) {
	logger := GetLogger()

	snapHelper(t, &logger, func() {
		logger.Verbose("this text should not be printed by a logger with default config")
	})

	snapHelper(t, &logger, func() {
		logger.VerboseEnabled = true
		logger.Verbose("hello")
	})
}

func TestLoggerWarning(t *testing.T) {
	logger := GetLogger()

	snapHelper(t, &logger, func() {
		logger.Warning("this text should be printed by a logger with default config %s", "(and in color)")
	})

	snapHelper(t, &logger, func() {
		logger.NoColor = true
		logger.Warning("this text should be printed by a logger with default config %s", "(but not be colorized)")
	})
}

func TestLoggerOutput(t *testing.T) {
	logger := GetLogger()

	snapHelper(t, &logger, func() {
		logger.Output("plain output should be printed always %s", "(also supporting format strings)")
	})
}

func TestLoggerError(t *testing.T) {
	logger := GetLogger()
	snapHelper(t, &logger, func() {
		logger.Error("this text should be printed by a logger with default config %s", "(and in color)")
	})

	snapHelper(t, &logger, func() {
		logger.NoColor = true
		logger.Error("this text should be printed by a logger with default config %s", "(but not be colorized)")
	})
}

func TestConfigure(t *testing.T) {
	modifiableLogger := Logger{
		VerboseEnabled: false,
		DebugEnabled:   false,
		NoColor:        false,
		writer:         nil,
	}

	if modifiableLogger.VerboseEnabled {
		t.Errorf("Assumption broken: VerboseEnabled was true already")
	}
	if modifiableLogger.DebugEnabled {
		t.Errorf("Assumption broken: DebugEnabled was true already")
	}
	if modifiableLogger.NoColor {
		t.Errorf("Assumption broken: NoColor was true already")
	}
	if modifiableLogger.writer != nil {
		t.Errorf("Assumption broken: writer was set already")
	}

	configLogger := Logger{
		VerboseEnabled: true,
		DebugEnabled:   true,
		NoColor:        true,
		writer:         os.Stderr,
	}

	modifiableLogger.Configure(configLogger)

	if !modifiableLogger.VerboseEnabled {
		t.Errorf("Configuring a logger with another logger did not lead to VerboseEnabled becoming true")
	}
	if !modifiableLogger.DebugEnabled {
		t.Errorf("Configuring a logger with another logger did not lead to DebugEnabled becoming true")
	}
	if !modifiableLogger.NoColor {
		t.Errorf("Configuring a logger with another logger did not lead to NoColor becoming true")
	}
	if modifiableLogger.writer != os.Stderr {
		t.Errorf("Configuring a logger with another logger did not lead to writer becoming set to os.Stderr")
	}
}

func TestMain(m *testing.M) {
	v := m.Run()

	// After all tests have run `go-snaps` will sort snapshots
	snaps.Clean(m, snaps.CleanOpts{Sort: true})

	os.Exit(v)
}
