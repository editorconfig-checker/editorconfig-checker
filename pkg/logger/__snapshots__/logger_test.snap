
[TestLoggerDebug - 1]

---

[TestLoggerDebug - 2]
this text should be printed when debug was enabled

---

[TestLoggerError - 1]
[31;1mthis text should be printed by a logger with default config (and in color)[33;0m

---

[TestLoggerError - 2]
this text should be printed by a logger with default config (but not be colorized)

---

[TestLoggerOutput - 1]
plain output should be printed always (also supporting format strings)

---

[TestLoggerVerbose - 1]

---

[TestLoggerVerbose - 2]
hello

---

[TestLoggerWarning - 1]
[33;1mthis text should be printed by a logger with default config (and in color)[33;0m

---

[TestLoggerWarning - 2]
this text should be printed by a logger with default config (but not be colorized)

---
